### PR TITLE
Pipeline buffering and upload in S3 multi-upload

### PR DESCRIPTION
**Pull request by losipiuk**
Rework of: https://github.com/trinodb/trino/pull/10180

Original PR desc:

> From discussion in https://github.com/trinodb/trino/pull/6201/files#r762349090, we should only wait for previous upload to finish when we need to flush the buffer.
> 
> Also, adding two minor nits:
> 1. set `failed` to `true` when skip multipart upload and directly upload but failed
> 2. Avoid allocating a new buffer when `finished` is `true`

It turned out that this PR results in corrupted writes to S3 (https://github.com/trinodb/trino/issues/10710).
The problem came from the fact that with `waitForPreviousUploadFinish` call moved from the very beginning of `flushBuffer` method further down, the check:
```
// skip multipart upload if there would only be one part
            if (finished && uploadId.isEmpty()) {
```
was no longer valid.
One call to `flushBuffer` could initiate the multipart upload, but the next call to `flushBuffer` could still observe `uploadId` as not set (there was no memory barrier between setting and reading threads).
Thanks @findepi for chat on this.

Co-authored-by: Łukasz Osipiuk <lukasz@osipuk.net>

Fixes https://github.com/trinodb/trino/issues/10710

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java [hunk 1]
```diff
@@ -1604,9 +1603,24 @@ private void flushBuffer(boolean finished)
             if (bufferSize == buffer.length || (finished && bufferSize > 0)) {
                 byte[] data = buffer;
                 int length = bufferSize;
-                this.buffer = new byte[buffer.length];
-                bufferSize = 0;
 
+                if (finished) {
+                    this.buffer = null;
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1608]: looks like unrelated optimization, move to separate commit

**Review by findepi — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java (hunk 1 above)
Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1608]: Done. 
Also added a tiny `Make upload abort idempotent` commit.

**Comment by losipiuk** [comment]
AC

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java [hunk 2]
```diff
@@ -1481,6 +1481,7 @@ public synchronized void progressChanged(ProgressEvent progressEvent)
         private Optional<String> uploadId = Optional.empty();
         private Future<UploadPartResult> inProgressUploadFuture;
         private final List<UploadPartResult> parts = new ArrayList<>();
+        private boolean aborted;
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1485]: is `aborted` much different than `failed`?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java [hunk 3]
```diff
@@ -1670,6 +1671,10 @@ private void finishUpload(String uploadId)
 
         private void abortUpload()
         {
+            if (aborted) {
+                return;
+            }
+            aborted = true;
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1684]: should `failed` be true now too?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java [hunk 4]
```diff
@@ -1478,6 +1478,7 @@ public synchronized void progressChanged(ProgressEvent progressEvent)
         private int bufferSize;
 
         private boolean failed;
+        private boolean multipartUploadStarted;
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1481]: there are effectively 2 threads here

- calling thread (one thread; as it's not meant to be thread-safe)
- the background uploading thread (because we never submit another task to executor until previous finished)

let's document, for each field, which thread(s) use it

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java [hunk 5]
```diff
@@ -1572,17 +1574,11 @@ public void close()
...
-                failed = true;
-                abortUploadSuppressed(e);
-                throw e;
-            }
+            // Skip multipart upload if there would only be one part
 
-            // skip multipart upload if there would only be one part
-            if (finished && uploadId.isEmpty()) {
+            // We use multipartUploadStarted flag here instead  uploadId.ifPresent() because visibility of latter one is not
+            // ensured here. It is mutated in other thread spawned by uploadExecutor.submit below
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1580]: This comment is understandable from reviewer and historical perspective. 
We wouldn't write such comment if we had the code right from the start, which suggests the comment is more suitable for the commit message than the code.

**Review by linzebing — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java (hunk 3 above)
Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1684]: will drop `abort` change for now to keep this PR focused.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java (hunk 4 above)
Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1482]: done.

And dropped other commits, replacing with commit message
